stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/sync/internal/SyncManager.kt

```diff
@@ -245,86 +250,102 @@ internal class SyncManager(
+            val message = repos.selectMessage(id) ?: continue
             val channelClient = chatClient.channel(message.cid)
-
-            when {
+            val result = when {
                 message.deletedAt != null -> {
-                    logger.logD("Deleting message: ${message.id}")
+                    logger.logV("[retryMgsWithSyncedAttachments] deleting message($id)")
                     channelClient.deleteMessage(message.id).await()
                 }
                 message.updatedLocallyAt != null && message.createdAt != null -> {
-                    logger.logD("Updating message: ${message.id}")
-
+                    logger.logV("[retryMgsWithSyncedAttachments] updating message($id)")
                     channelClient.updateMessage(message).await()
                 }
                 else -> {
-                    logger.logD("Sending message: ${message.id}")
-                    val result = channelClient.sendMessage(message).await()
-
-                    if (result.isSuccess) {
-                        repos.insertMessage(message.copy(syncStatus = SyncStatus.COMPLETED))
-                    } else if (result.isError && result.error().isPermanent()) {
-                        markMessageAsFailed(message)
+                    logger.logV("[retryMgsWithSyncedAttachments] sending message($id)")
+                    channelClient.sendMessage(message).await().also { result ->
+                        if (result.isSuccess) {
+                            repos.insertMessage(message.copy(syncStatus = SyncStatus.COMPLETED))
+                        } else if (result.isError && result.error().isPermanent()) {
+                            repos.markMessageAsFailed(message)
+                        }
                     }
                 }
             }
+            logger.logV("[retryMgsWithSyncedAttachments] result($id).isSuccess: ${result.isSuccess}")
         }
-
-        return messages
     }
 
     /**
      * Retries messages with [SyncStatus.AWAITING_ATTACHMENTS] status.
      */
-    private suspend fun retryMessagesWithPendingAttachments(): List<Message> {
-        val retriedMessages = repos.selectMessageBySyncState(SyncStatus.AWAITING_ATTACHMENTS)
-
-        val (failedMessages, needToBeSync) = retriedMessages.partition { message ->
-            message.attachments.any { it.uploadState is Attachment.UploadState.Failed }
-        }
-
-        failedMessages.forEach { markMessageAsFailed(it) }
-
-        needToBeSync.forEach { message ->
-            val (channelType, channelId) = message.cid.cidToTypeAndId()
-            chatClient.sendMessage(channelType, channelId, message, true).await()
+    private suspend fun retryMessagesWithPendingAttachments() {
+        val ids = repos.selectMessageIdsBySyncState(SyncStatus.AWAITING_ATTACHMENTS)
+        logger.logD("[retryMessagesWithPendingAttachments] ids.size: ${ids.size}")
+        for (id in ids) {
```

Comment:
Overall we don't use `for` loops. I believe you could use `fold` in this case to merge all the messages into a list and then iterate over them.